terrain_simplification/src/terrain_simplification.cpp

```diff
@@ -0,0 +1,258 @@
+#include <terrain_simplification/terrain_simplification.hpp>
+
+#include <opencv2/core/core.hpp>
+#include <opencv2/imgcodecs.hpp>
+#include <opencv2/highgui/highgui.hpp>
+#include <grid_map_cv/GridMapCvConverter.hpp>
+
+#include <ros/ros.h>
+#include <iostream>
+
+using Clock = std::chrono::high_resolution_clock;
+
+namespace terrain_simplification {
+
+TerrainSimplification::TerrainSimplification ()
+  : thread_loop_(true),
+    thread_run_(&TerrainSimplification::run, this) {
+}
+
+TerrainSimplification::~TerrainSimplification() {
+  if (thread_run_.joinable()) {
+    thread_loop_ = false;
+    // Attempting to join the thread running the simplification with the main thread of TerrainSimplification...
+    thread_run_.join();
+  } else {
+    std::cerr << "[TerrainSimplification::~TerrainSimplification] "
+                 "The thread running the MPC could NOT be joined with "
+                 "the main thread of TerrainSimplification: "
+                 "thread_run_.joinable() = " << std::endl;
+  }
+}
+
+void
+TerrainSimplification::run() {
+  // continously run the thread, until stopped
+  while (thread_loop_.load()) {
+    // IF flags true, advance the simplifying algorithm to produce to create simplified map
+    // ELSE wait for the flags
+    if (run_ && received_) {
+      advance();
+    } else {
+      // wait 0.2 seconds
+      for (int i = 0; i < 100; i++) {
+        usleep(2000);
+      }
+    }
+  }
+}
+
+bool
+TerrainSimplification::advance() {
+  simplifyGridMap();
+  return true;
+}
+
+void
+TerrainSimplification::setGridMap(
+    const grid_map::GridMap& map) {
+  map_full_ = map;
+  received_ = true;
+}
+
+double
+TerrainSimplification::getRobotYaw() {
+  mutex_state_.lock();
```

Comment:
Instead of locking and unlocking at the end of a method, you can use this : http://www.cplusplus.com/reference/mutex/lock_guard/